Fix `test_mulled_build.py::test_mulled_build_files_cli` with `use_mamba=True`

domgz opened the pull request:
Just a quick attempt at reducing the amount of failing tests, feel free to step in with a better solution.

---

Test `tests/tool_util/mulled/test_mulled_build.py::test_mulled_build_files_cli[True]`, where `[True]` refers to the parameter `use_mamba`, fails because if `conda install --quiet --yes 'mamba='` (the preinstall command) runs before `mamba install -p /usr/local`, then the latter expects either `/usr/local` not to exist or to be an existing environment.

To work this around, create an environment in `/usr/local/env`, but still put it on the expected location `/usr/local` later.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

domgz committed:
Fix `test_mulled_build.py::test_mulled_build_files_cli` with `use_mamba=True`

Test `tests/tool_util/mulled/test_mulled_build.py::test_mulled_build_files_cli[True]`, where `[True]` refers to the parameter `use_mamba`, fails because if `conda install --quiet --yes 'mamba='` (the preinstall command) runs before `mamba install -p /usr/local`, then the latter expects either `/usr/local` not to exist or to be an existing environment.

To work this around, create an environment in `/usr/local/env`, but still put it on the expected location `/usr/local` later.

nsoranzo commented on lib/galaxy/tool_util/deps/mulled/invfile.lua:
```diff
@@ -88,13 +88,14 @@ inv.task('build')
     .using(conda_image)
         .withHostConfig({binds = bind_args})
         .run('/bin/sh', '-c', preinstall
+            .. conda_bin .. ' create --quiet --yes -p /usr/local/env --copy  && '
```
Maybe execute this as a separate command to make future debugging easier?

domgz replied:
Do you mean not using `&&`, wrapping it in another `.run()` statement, or something else?

nsoranzo replied:
Yes, in a separate `run()`, if that makes sense.

domgz replied:
I do not know if the whole thing is designed to work this way. [According to the docs](https://github.com/involucro/involucro/blob/master/README.md#run-step), every `run()` statement runs in its own container.

For example, this invfile that creates a file `/file.txt` and then attempts to read it

```lua
inv.task('build')
    .using(conda_image)
        .withHostConfig({binds = {"build:/data"}})
        .run('rm', '-rf', '/data/dist')
    .using(conda_image)
        .withHostConfig({binds = bind_args})
        .run('/bin/sh', '-c', preinstall)
        .run('/bin/sh', '-c', conda_bin .. ' create --quiet --yes -p /usr/local/env --copy')
        .run('/bin/sh', '-c', conda_bin .. ' install ' .. channel_args .. ' ' .. target_args
             .. ' --strict-channel-priority -p /usr/local/env --copy --yes ' .. verbose)
        .run('/bin/sh', '-c', postinstall)
        .run('/bin/sh', '-c', 'echo abc > /file.txt')
        .run('/bin/sh', '-c', 'echo def > /file.txt')
        .run('/bin/sh', '-c', 'cat /file.txt')
    .wrap('build/dist/env')
        .at('/usr/local')
        .inImage(destination_base_image)
        .as(repo)
```

leads to a failure reading the file when running `test_mulled_build.py::test_mulled_build_files_cli[True]`

```
use_mamba = True, tmpdir = local('/tmp/pytest-of-jose/pytest-0/test_mulled_build_files_cli_Tr0')

    @pytest.mark.parametrize("use_mamba", [False, True])
    @external_dependency_management
    def test_mulled_build_files_cli(use_mamba: bool, tmpdir) -> None:
        singularity_image_dir = tmpdir.mkdir("singularity image dir")
        target = build_target("zlib", version="1.2.13", build="h166bdaf_4")
        involucro_context = InvolucroContext(involucro_bin=os.path.join(tmpdir, "involucro"))
        exit_code = mull_targets(
            [target],
            involucro_context=involucro_context,
            command="build-and-test",
            singularity=True,
            use_mamba=use_mamba,
            singularity_image_dir=singularity_image_dir,
        )
>       assert exit_code == 0
E       assert 1 == 0

exit_code  = 1
involucro_context = <galaxy.tool_util.deps.mulled.mulled_build.InvolucroContext object at 0x70093661e520>
singularity_image_dir = local('/tmp/pytest-of-jose/pytest-0/test_mulled_build_files_cli_Tr0/singularity image dir')
target     = CondaTarget[package=zlib,version=1.2.13,build=h166bdaf_4]
tmpdir     = local('/tmp/pytest-of-jose/pytest-0/test_mulled_build_files_cli_Tr0')
use_mamba  = True

test/unit/tool_util/mulled/test_mulled_build.py:47: AssertionError

----------------------------- Captured stdout call -----------------------------
Executing: /tmp/pytest-of-jose/pytest-0/test_mulled_build_files_cli_Tr0/involucro -v=3 -f /home/jose/Dokumente/Repositories/galaxy/lib/galaxy/tool_util/deps/mulled/invfile.lua -set CHANNELS=conda-forge,bioconda -set TARGETS=zlib=1.2.13=h166bdaf_4 -set REPO=quay.io/biocontainers/zlib:1.2.13--h166bdaf_4 -set BINDS=build/dist:/usr/local/ -set DEST_BASE_IMAGE=quay.io/bioconda/base-glibc-busybox-bash:latest -set CONDA_IMAGE=quay.io/condaforge/miniforge3:latest -set SINGULARITY=1 -set SINGULARITY_IMAGE_NAME=zlib:1.2.13--h166bdaf_4 -set 'SINGULARITY_IMAGE_DIR=/tmp/pytest-of-jose/pytest-0/test_mulled_build_files_cli_Tr0/singularity image dir' -set USER_ID=1000:1000 -set TEST=true -set CONDA_BIN=mamba -set 'PREINSTALL=if true ; then conda install --quiet --yes '"'"'mamba='"'"' ; fi' build-and-test
..

----------------------------- Captured stderr call -----------------------------
[Feb  6 12:09:49] DEBU Run file [/home/jose/Dokumente/Repositories/galaxy/lib/galaxy/tool_util/deps/mulled/invfile.lua]
[Feb  6 12:09:49] INFO Invoke Task [build]
[Feb  6 12:09:49] STEP Run image [quay.io/condaforge/miniforge3:latest] with command [[rm -rf /data/dist]]
[Feb  6 12:09:49] DEBU Creating container [step-7daa022bfc]
[Feb  6 12:09:49] DEBU Created container [f676d194dcc6 step-7daa022bfc], starting it
[Feb  6 12:09:49] DEBU Container [f676d194dcc6 step-7daa022bfc] started, waiting for completion
[Feb  6 12:09:49] DEBU Container [f676d194dcc6 step-7daa022bfc] completed with exit code [0] as expected
[Feb  6 12:09:49] DEBU Container [f676d194dcc6 step-7daa022bfc] removed
[Feb  6 12:09:49] STEP Run image [quay.io/condaforge/miniforge3:latest] with command [[/bin/sh -c if true ; then conda install --quiet --yes 'mamba=' ; fi]]
[Feb  6 12:09:49] DEBU Creating container [step-d20b95e90c]
[Feb  6 12:09:49] DEBU Created container [becb2d5a7eea step-d20b95e90c], starting it
[Feb  6 12:09:49] DEBU Container [becb2d5a7eea step-d20b95e90c] started, waiting for completion
[Feb  6 12:09:50] SOUT Channels:
[Feb  6 12:09:50] SOUT - conda-forge
[Feb  6 12:09:50] SOUT Platform: linux-64
[Feb  6 12:10:17] SOUT Collecting package metadata (repodata.json): ...working... done
[Feb  6 12:10:18] SOUT Solving environment: ...working... done
[Feb  6 12:10:19] SOUT 
[Feb  6 12:10:19] SOUT ## Package Plan ##
[Feb  6 12:10:19] SOUT 
[Feb  6 12:10:19] SOUT environment location: /opt/conda
[Feb  6 12:10:19] SOUT 
[Feb  6 12:10:19] SOUT added / updated specs:
[Feb  6 12:10:19] SOUT - mamba
[Feb  6 12:10:19] SOUT 
[Feb  6 12:10:19] SOUT 
[Feb  6 12:10:19] SOUT The following packages will be downloaded:
[Feb  6 12:10:19] SOUT 
[Feb  6 12:10:19] SOUT package                    |            build
[Feb  6 12:10:19] SOUT ---------------------------|-----------------
[Feb  6 12:10:19] SOUT ca-certificates-2025.1.31  |       hbcca054_0         154 KB  conda-forge
[Feb  6 12:10:19] SOUT conda-25.1.1               |  py312h7900ff3_0         1.1 MB  conda-forge
[Feb  6 12:10:19] SOUT conda-libmamba-solver-25.1.1|     pyhd8ed1ab_0          40 KB  conda-forge
[Feb  6 12:10:19] SOUT cpp-expected-1.1.0         |       hf52228f_0          23 KB  conda-forge
[Feb  6 12:10:19] SOUT libmamba-2.0.5             |       h49b8a8d_1         1.9 MB  conda-forge
[Feb  6 12:10:19] SOUT libmambapy-2.0.5           |  py312hbaee817_1         609 KB  conda-forge
[Feb  6 12:10:19] SOUT mamba-2.0.5                |       h8871ed4_1         463 KB  conda-forge
[Feb  6 12:10:19] SOUT nlohmann_json-3.11.3       |       he02047a_1         120 KB  conda-forge
[Feb  6 12:10:19] SOUT simdjson-3.11.6            |       h84d6215_0         240 KB  conda-forge
[Feb  6 12:10:19] SOUT spdlog-1.15.1              |       hb29a8c4_0         190 KB  conda-forge
[Feb  6 12:10:19] SOUT ------------------------------------------------------------
[Feb  6 12:10:19] SOUT Total:         4.8 MB
[Feb  6 12:10:19] SOUT 
[Feb  6 12:10:19] SOUT The following NEW packages will be INSTALLED:
[Feb  6 12:10:19] SOUT 
[Feb  6 12:10:19] SOUT cpp-expected       conda-forge/linux-64::cpp-expected-1.1.0-hf52228f_0
[Feb  6 12:10:19] SOUT nlohmann_json      conda-forge/linux-64::nlohmann_json-3.11.3-he02047a_1
[Feb  6 12:10:19] SOUT simdjson           conda-forge/linux-64::simdjson-3.11.6-h84d6215_0
[Feb  6 12:10:19] SOUT spdlog             conda-forge/linux-64::spdlog-1.15.1-hb29a8c4_0
[Feb  6 12:10:19] SOUT 
[Feb  6 12:10:19] SOUT The following packages will be UPDATED:
[Feb  6 12:10:19] SOUT 
[Feb  6 12:10:19] SOUT ca-certificates                     2024.12.14-hbcca054_0 --> 2025.1.31-hbcca054_0
[Feb  6 12:10:19] SOUT conda                             24.11.3-py312h7900ff3_0 --> 25.1.1-py312h7900ff3_0
[Feb  6 12:10:19] SOUT conda-libmamba-so~                    24.9.0-pyhd8ed1ab_0 --> 25.1.1-pyhd8ed1ab_0
[Feb  6 12:10:19] SOUT libmamba                                1.5.12-h49b8a8d_0 --> 2.0.5-h49b8a8d_1
[Feb  6 12:10:19] SOUT libmambapy                         1.5.12-py312hbaee817_0 --> 2.0.5-py312hbaee817_1
[Feb  6 12:10:19] SOUT mamba                              1.5.12-py312h9460a1c_0 --> 2.0.5-h8871ed4_1
[Feb  6 12:10:19] SOUT 
[Feb  6 12:10:19] SOUT 
[Feb  6 12:10:19] SOUT Preparing transaction: ...working... done
[Feb  6 12:10:19] SOUT Verifying transaction: ...working... done
[Feb  6 12:10:20] SOUT Executing transaction: ...working... done
[Feb  6 12:10:21] DEBU Container [becb2d5a7eea step-d20b95e90c] completed with exit code [0] as expected
[Feb  6 12:10:22] DEBU Container [becb2d5a7eea step-d20b95e90c] removed
[Feb  6 12:10:22] STEP Run image [quay.io/condaforge/miniforge3:latest] with command [[/bin/sh -c mamba create --quiet --yes -p /usr/local/env --copy]]
[Feb  6 12:10:22] DEBU Creating container [step-06e635be44]
[Feb  6 12:10:22] DEBU Created container [f5fc06174a30 step-06e635be44], starting it
[Feb  6 12:10:22] DEBU Container [f5fc06174a30 step-06e635be44] started, waiting for completion
[Feb  6 12:10:23] SOUT Preparing transaction: ...working... done
[Feb  6 12:10:23] SOUT Verifying transaction: ...working... done
[Feb  6 12:10:23] SOUT Executing transaction: ...working... done
[Feb  6 12:10:23] DEBU Container [f5fc06174a30 step-06e635be44] completed with exit code [0] as expected
[Feb  6 12:10:23] DEBU Container [f5fc06174a30 step-06e635be44] removed
[Feb  6 12:10:23] STEP Run image [quay.io/condaforge/miniforge3:latest] with command [[/bin/sh -c mamba install  -c 'conda-forge' -c 'bioconda'  'zlib=1.2.13=h166bdaf_4' --strict-channel-priority -p /usr/local/env --copy --yes --quiet]]
[Feb  6 12:10:23] DEBU Creating container [step-6926ee7168]
[Feb  6 12:10:23] DEBU Created container [6f56847bd91c step-6926ee7168], starting it
[Feb  6 12:10:23] DEBU Container [6f56847bd91c step-6926ee7168] started, waiting for completion
[Feb  6 12:10:55] SOUT Preparing transaction: ...working... done
[Feb  6 12:10:55] SOUT Verifying transaction: ...working... done
[Feb  6 12:10:55] SOUT Executing transaction: ...working... done
[Feb  6 12:10:58] DEBU Container [6f56847bd91c step-6926ee7168] completed with exit code [0] as expected
[Feb  6 12:10:58] DEBU Container [6f56847bd91c step-6926ee7168] removed
[Feb  6 12:10:58] STEP Run image [quay.io/condaforge/miniforge3:latest] with command [[/bin/sh -c ]]
[Feb  6 12:10:58] DEBU Creating container [step-a176ea7391]
[Feb  6 12:10:59] DEBU Created container [362a5821c3e9 step-a176ea7391], starting it
[Feb  6 12:10:59] DEBU Container [362a5821c3e9 step-a176ea7391] started, waiting for completion
[Feb  6 12:10:59] DEBU Container [362a5821c3e9 step-a176ea7391] completed with exit code [0] as expected
[Feb  6 12:10:59] DEBU Container [362a5821c3e9 step-a176ea7391] removed
[Feb  6 12:10:59] STEP Run image [quay.io/condaforge/miniforge3:latest] with command [[/bin/sh -c echo abc > /file.txt]]
[Feb  6 12:10:59] DEBU Creating container [step-1bb105e474]
[Feb  6 12:10:59] DEBU Created container [44d4824a1113 step-1bb105e474], starting it
[Feb  6 12:10:59] DEBU Container [44d4824a1113 step-1bb105e474] started, waiting for completion
[Feb  6 12:10:59] DEBU Container [44d4824a1113 step-1bb105e474] completed with exit code [0] as expected
[Feb  6 12:10:59] DEBU Container [44d4824a1113 step-1bb105e474] removed
[Feb  6 12:10:59] STEP Run image [quay.io/condaforge/miniforge3:latest] with command [[/bin/sh -c echo def > /file.txt]]
[Feb  6 12:10:59] DEBU Creating container [step-3bce72b19f]
[Feb  6 12:10:59] DEBU Created container [766e6b994180 step-3bce72b19f], starting it
[Feb  6 12:10:59] DEBU Container [766e6b994180 step-3bce72b19f] started, waiting for completion
[Feb  6 12:10:59] DEBU Container [766e6b994180 step-3bce72b19f] completed with exit code [0] as expected
[Feb  6 12:10:59] DEBU Container [766e6b994180 step-3bce72b19f] removed
[Feb  6 12:10:59] STEP Run image [quay.io/condaforge/miniforge3:latest] with command [[/bin/sh -c cat /file.txt]]
[Feb  6 12:10:59] DEBU Creating container [step-3ce28e3247]
[Feb  6 12:10:59] DEBU Created container [a659353c69c4 step-3ce28e3247], starting it
[Feb  6 12:11:00] DEBU Container [a659353c69c4 step-3ce28e3247] started, waiting for completion
[Feb  6 12:11:00] SERR cat: /file.txt: No such file or directory
[Feb  6 12:11:00] ERRO Task processing failed: Unexpected exit code [1] of container [a659353c69c4 step-3ce28e3247], container preserved

------------------------------ Captured log call -------------------------------
DEBUG    urllib3.connectionpool:connectionpool.py:291 Resetting dropped connection: conda.anaconda.org
DEBUG    urllib3.connectionpool:connectionpool.py:475 https://conda.anaconda.org:443 "GET /conda-forge/linux-64/zlib-1.2.13-h166bdaf_4.tar.bz2 HTTP/1.1" 200 94099
DEBUG    urllib3.connectionpool:connectionpool.py:1022 Starting new HTTPS connection (1): github.com:443
DEBUG    urllib3.connectionpool:connectionpool.py:475 https://github.com:443 "GET /involucro/involucro/releases/download/v1.1.2/involucro HTTP/1.1" 302 0
DEBUG    urllib3.connectionpool:connectionpool.py:1022 Starting new HTTPS connection (1): objects.githubusercontent.com:443
DEBUG    urllib3.connectionpool:connectionpool.py:475 https://objects.githubusercontent.com:443 "GET /github-production-release-asset-2e65be/46847994/7a4a2352-6f82-11e6-8e67-14ed4f0d047a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250206%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250206T110812Z&X-Amz-Expires=300&X-Amz-Signature=620a0c2af6a675b15af67f5e174c46e4830a674f51dd70d49b04320b1c0f7314&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dinvolucro&response-content-type=application%2Foctet-stream HTTP/1.1" 200 8112832
```

Given the way it currently works, it cannot be ruled out that other code is building the preinstall or postinstall commands under the assumption that they will run in the same container (or even in the same container as the conda commands).

My intenton is just to fix the CI, and several `run()` statements can potentially introduce new bugs :sweat_smile:

domgz replied:
I did nevertheless commit bd24d0c, but maybe we should revert it

nsoranzo replied:
The docs are bit unclear, but your test surely shows that my idea doesn't work, so please revert the commit. Thanks a lot for trying though!

domgz committed:
Split installation shell command into four

With the aim of easing debugging, write one `run()` statement for the preinstall command, one to create the conda environment, another to install packages to the environment and one for the postinstall command.

domgz committed:
Revert "Split installation shell command into four"

This reverts commit bd24d0cafc14ba347e9a50713ebece6b630e7dc1.

It makes sense to revert the changes because every `run()` statement runs in its own container, thus it cannot be ruled out that other code is building the preinstall or postinstall commands under the assumption that they will run in the same container (or even in the same container as the conda commands). An example is available on https://github.com/galaxyproject/galaxy/pull/19545#discussion_r1944557524.
